### PR TITLE
EASYOPAC-942 - Implemented changes and fixes.

### PR DESCRIPTION
**Pull request by bozeak**

### CODE REVIEW

File: modules/opensearch/opensearch.client.inc [hunk 1]
```diff
@@ -376,7 +376,7 @@ function opensearch_execute_cache($request) {
 
   // Handle fulltext vs. dkabm caching of object.
   $type = OPENSEARCH_CACHE_TING_OBJECT;
-  if ($parms['objectFormat'] == 'docbook') {
+  if ($parms['format'] == 'docbook') {
```
Comment (guddo) [inline modules/opensearch/opensearch.client.inc]: Why did you change this?

File: modules/ting/ting.entities.inc [hunk 2]
```diff
@@ -298,7 +298,19 @@ class TingEntity extends DingEntity {
    *   The language, FALSE if it could not be found.
    */
   public function getLanguage() {
-    return empty($this->reply) ? FALSE : $this->reply->getLanguage();
```
Comment (guddo) [inline modules/ting/ting.entities.inc]: Are you sure this does not belong in https://github.com/easySuite/ding2/blob/development-4.4.x/modules/opensearch/src/OpenSearchTingObject.php#L202?

File: modules/ting/ting.entities.inc [hunk 3]
```diff
@@ -352,7 +364,17 @@ class TingEntity extends DingEntity {
    *   The description.
    */
   public function getDescription() {
-    return empty($this->reply) ? FALSE : $this->reply->getDescription();
+    $description = FALSE;
```
Comment (guddo) [inline modules/ting/ting.entities.inc]: Are you sure this does not belong in https://github.com/easySuite/ding2/blob/development-4.4.x/modules/opensearch/src/OpenSearchTingObject.php#L393?

File: modules/ting/ting.entities.inc [hunk 4]
```diff
@@ -366,6 +388,18 @@ class TingEntity extends DingEntity {
   public function getIsbn() {
     return empty($this->reply) ? FALSE : $this->reply->getIsbn();
   }
+
+  /**
+   * The Classification Text.
```
Comment (guddo) [inline modules/ting/ting.entities.inc]: Are you sure this should not go into https://github.com/easySuite/ding2/blob/development-4.4.x/modules/opensearch/src/OpenSearchTingObject.php#L161?

File: modules/opensearch/opensearch.client.inc (hunk 1 above)
Comment (bozeak) [reply modules/opensearch/opensearch.client.inc]: Because in all situations 'objectFormat' key was completely missing.

**Comment by bozeak** [comment]
> Does ting_search_disable_type work as it should?

Yes, it is working as it is expected.

> What about changes in EASYOPAC-837 and EASYOPAC-839?

Also was checked and it was ok.

**Comment by guddo** [comment]
>> What about changes in EASYOPAC-837 and EASYOPAC-839?

> Also was checked and it was ok.

I understand, but are they really needed in this new version? Did you test with and without these patches on user pages with ILL and non-ILL loans?

File: modules/ting/ting.entities.inc (hunk 2 above)
Comment (guddo) [reply modules/ting/ting.entities.inc]: Please relate to my comment.

File: modules/ting/ting.entities.inc (hunk 3 above)
Comment (guddo) [reply modules/ting/ting.entities.inc]: Please relate to my comment.

File: modules/ting/ting.entities.inc (hunk 4 above)
Comment (guddo) [reply modules/ting/ting.entities.inc]: Please relate to my comment.

File: modules/opensearch/opensearch.client.inc (hunk 1 above)
Comment (guddo) [reply modules/opensearch/opensearch.client.inc]: Does this change relate to any of the mentioned elements in the case?

Comment (bozeak) [reply modules/opensearch/opensearch.client.inc]: No, directly it is not related. Only reason why this was changed is that this incorrect/missing key was providing annoying notices on front page and in dblog.

File: modules/ting/ting.entities.inc (hunk 3 above)
Comment (bozeak) [reply modules/ting/ting.entities.inc]: In https://github.com/easySuite/ding2/blob/development-4.4.x/modules/opensearch/src/OpenSearchTingObject.php#L393 will be outputted all text found in description. Code proposed to be added will trim description value to predefined amount of symbols. Don't we need such functionality?

File: modules/ting/ting.entities.inc (hunk 2 above)
Comment (guddo) [reply modules/ting/ting.entities.inc]: I mean; are you sure that the code you have added should be added in `ting/ting.entities.inc` and not in `opensearch/src/OpenSearchTingObject.php` instead?

Comment (bozeak) [reply modules/ting/ting.entities.inc]: No. Because in a final we are processing TingEntity object which is going through TingEntity class declared in ting/ting.entities.inc.

File: modules/ting/ting.entities.inc (hunk 3 above)
Comment (bozeak) [reply modules/ting/ting.entities.inc]: No. Because in a final we are processing TingEntity object which is going through TingEntity class declared in ting/ting.entities.inc.

File: modules/ting/ting.entities.inc (hunk 4 above)
Comment (bozeak) [reply modules/ting/ting.entities.inc]: No. Because in a final we are processing TingEntity object which is going through TingEntity class declared in ting/ting.entities.inc.

**Comment by bozeak** [comment]
> I understand, but are they really needed in this new version? Did you test with and without these patches on user pages with ILL and non-ILL loans?

How can I test without patches while those was merged long time ago in development branch?

File: modules/opensearch/opensearch.client.inc (hunk 1 above)
Comment (guddo) [reply modules/opensearch/opensearch.client.inc]: Ok, lets revert this for now as it does not relate to the issues this case should resolve.

**Comment by guddo** [comment]
Nevermind, lets keep discussio about EASYOPAC-837 in https://github.com/easySuite/ding2/pull/493#pullrequestreview-161547827.

File: modules/opensearch/src/OpenSearchTingObject.php [hunk 5]
```diff
@@ -200,7 +213,19 @@ public function getSubjects() {
    * {@inheritdoc}
    */
   public function getLanguage() {
-    return $this->firstEntry($this->getRecordEntry('dc:language'));
+    if (!isset($this->reply->record['dc:language'])) {
```
Comment (guddo) [inline modules/opensearch/src/OpenSearchTingObject.php]: Shouldn't this be `$this->firstEntry($this->getRecordEntry)`?

Comment (bozeak) [reply modules/opensearch/src/OpenSearchTingObject.php]: `$this->firstEntry($this->getRecordEntry('dc:language'))` will return plain string, but in case there are multiple languages we need an array for processing and returning the list of languages.

Comment (bozeak) [reply modules/opensearch/src/OpenSearchTingObject.php]: > Are you sure about this? I'm quite sure that if there is multiple languages the language code is just "mul" instead of "da", "en", etc.

No. In case when language code is "mul", the "mul" is the first element, next are listed other languages available.
Example: https://oss-services.dbc.dk/opensearch/5.0/?action=search&query=%22Kvartalsskriftet%22&agency=100200&profile=test&start=1&stepValue=5

> If its an array, why is the variable name in L220 called $languages_string?

Because array values with language name are concatenated in one string I suppose.

Comment (guddo) [reply modules/opensearch/src/OpenSearchTingObject.php]: Ok, should it be `$this->getRecordEntry` instead then? I dont see `$this->reply` used anywhere else in this file.

Comment (bozeak) [reply modules/opensearch/src/OpenSearchTingObject.php]: In case of `$this->getRecordEntry` will be returned **"Flere sprog"** string and rest of items will be skipped.
